### PR TITLE
Enhance class SoftmaxRegression

### PR DESCRIPTION
**Pull request by stereomatchingkiss**
1 : add function Serialize to class SoftmaxRegression
2 : add function Train to class SoftmaxRegression
3 : fix bug--did not initialize fitIntercept

The test case "SoftmaxRegressionFunctionEvaluate" always fail on my machine

### CODE REVIEW

File: src/mlpack/methods/softmax_regression/softmax_regression_impl.hpp [hunk 1]
```diff
@@ -14,47 +14,59 @@ namespace mlpack {
...
 
 template<template<typename> class OptimizerType>
+SoftmaxRegression<OptimizerType>::
+SoftmaxRegression(const size_t inputSize,
+                  const size_t numClasses,
+                  const bool fitIntercept) :
+  inputSize{inputSize},
+  numClasses{numClasses},
+  lambda{0.0001},
+  fitIntercept{fitIntercept}
```
Comment (rcurtin) [inline src/mlpack/methods/softmax_regression/softmax_regression_impl.hpp]: For consistency, do you mind if I change this to `()` initialization vs. `{}` initialization?

File: src/mlpack/methods/softmax_regression/softmax_regression_impl.hpp [hunk 2]
```diff
@@ -14,47 +14,59 @@ namespace mlpack {
...
-    parameters(optimizer.Function().GetInitialPoint()),
-    inputSize(optimizer.Function().InputSize()),
-    numClasses(optimizer.Function().NumClasses()),
-    lambda(optimizer.Function().Lambda())
+  OptimizerType<SoftmaxRegressionFunction>& optimizer) :
+  parameters(optimizer.Function().GetInitialPoint()),
+  inputSize{optimizer.Function().InputSize()},
+  numClasses{optimizer.Function().NumClasses()},
+  lambda{optimizer.Function().Lambda()},
+  fitIntercept{optimizer.Function().FitIntercept()}
```
Comment (rcurtin) [inline src/mlpack/methods/softmax_regression/softmax_regression_impl.hpp]: This should be four spaces in from the previous line; I'll fix this when I merge it in.  (It happens a couple other places too.)  I'm just pointing it out here so that you know. :)

File: src/mlpack/methods/softmax_regression/softmax_regression.hpp [hunk 3]
```diff
@@ -57,10 +57,34 @@ namespace regression {
...
+  /**
+   * Construct the SoftmaxRegression class with the provided data and labels.
+   * This will train the model.
+   *
+   * @param fileName name of the files saving the model contents
+   * @param name name of the structure to be save
+   * @exception If the file cannot be load, the exception will thrown
+   */
+  SoftmaxRegression(const std::string &fileName,
+                    const std::string& name);
```
Comment (rcurtin) [inline src/mlpack/methods/softmax_regression/softmax_regression.hpp]: I think we should avoid this constructor, because no other mlpack methods provide a constructor of this sort.  If we implemented this method here, then we'd have to implement it anywhere (for consistency).  I'd prefer that users use `data::Load()` instead (maybe we can revisit this later, but for now I don't have time to create that constructor for every mlpack method before the next release).

**Comment by rcurtin** [comment]
Looks good to me; I made a few comments.  Let me know what you think.  I'll add a test for `SoftmaxRegression::Serialize()` once I merge.  Do you mind if I add your name/email to the list of authors?  If that's okay, which name/email should I use?

File: src/mlpack/methods/softmax_regression/softmax_regression_impl.hpp (hunk 1 above)
Comment (stereomatchingkiss) [inline src/mlpack/methods/softmax_regression/softmax_regression_impl.hpp]: No problem(I use {} to avoid narrow conversion)

File: src/mlpack/methods/softmax_regression/softmax_regression_impl.hpp (hunk 2 above)
Comment (stereomatchingkiss) [inline src/mlpack/methods/softmax_regression/softmax_regression_impl.hpp]: Sorry, my mistake, do you mind using some tools to do the formatting(ex : uncrustify)?

File: src/mlpack/methods/softmax_regression/softmax_regression.hpp (hunk 3 above)
Comment (stereomatchingkiss) [inline src/mlpack/methods/softmax_regression/softmax_regression.hpp]: Agree with you, consistency is very important

**Comment by stereomatchingkiss** [comment]
>Do you mind if I add your name/email to the list of authors? 
My pleasure

name : Tham Ngap Wei
email : thamngapwei@gmail.com

File: src/mlpack/methods/softmax_regression/softmax_regression_impl.hpp (hunk 2 above)
Comment (rcurtin) [inline src/mlpack/methods/softmax_regression/softmax_regression_impl.hpp]: Sure, if you can make a good uncrustify config we can put it somewhere useful.  Personally I just do the formatting by hand, so I'll fix this when I merge it in.

**Comment by rcurtin** [comment]
Okay, I made some simple changes in cbeb3ea..f85a9b2; let me know if you see anything wrong.  If you are interested, here are some future possibilities for refactoring/improving the SoftmaxRegression code further:

 * Storing inputSize and numClasses is, I think, redundant, and I think you can just get these from the parameters matrix.  So the constructors can be cleaned up a little (although numClasses will still be necessary in the constructor).

 * The type of label would be better as `arma::Row<size_t>` not `arma::vec`.

 * The `InitializeWeights()` function could be made static or moved elsewhere so that creating a `SoftmaxRegressionFunction` to initialize the weights in the non-training constructor isn't necessary.

 * A `softmax_regression_main.cpp` program could be written that could allow users to use softmax regression from the command line.

Don't feel obligated to do any of these -- I'm just tossing them out there as ideas. :)

**Comment by stereomatchingkiss** [comment]
Your suggestions are good, about

>Storing inputSize and numClasses is, I think, redundant
Some(or many) classes of mlpack always ask for the inputSize(feature size) from the users, I think they could be removed too, because many algorithms should be able to find out the inputSize from the training data


>Thanks for your contribution! :)
You are welcome:), thanks for the helps from community too
